cmd/bccatomicswap/main.go

```diff
@@ -0,0 +1,1133 @@
+	}
+	contractTx, complete, err := c.SignRawTransaction(unsignedContract)
+	if err != nil {
+		return nil, fmt.Errorf("signrawtransaction: %v", err)
+	}
+	if !complete {
+		return nil, errors.New("signrawtransaction: failed to completely sign contract transaction")
+	}
+
+	contractTxHash := contractTx.TxHash()
+
+	refundTx, refundFee, err := buildRefund(c, contract, contractTx, feePerKb, minFeePerKb)
+	if err != nil {
+		return nil, err
+	}
+
+	return &builtContract{
+		contract,
+		contractP2SH,
+		&contractTxHash,
+		contractTx,
+		contractFee,
+		refundTx,
+		refundFee,
+	}, nil
+}
+
+func buildRefund(c *rpc.Client, contract []byte, contractTx *wire.MsgTx, feePerKb, minFeePerKb btcutil.Amount) (
+	refundTx *wire.MsgTx, refundFee btcutil.Amount, err error) {
+
+	contractP2SH, err := btcutil.NewAddressScriptHash(contract, chainParams)
+	if err != nil {
+		return nil, 0, err
+	}
+	contractP2SHPkScript, err := txscript.PayToAddrScript(contractP2SH)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	contractTxHash := contractTx.TxHash()
+	contractOutPoint := wire.OutPoint{Hash: contractTxHash, Index: ^uint32(0)}
+	for i, o := range contractTx.TxOut {
+		if bytes.Equal(o.PkScript, contractP2SHPkScript) {
+			contractOutPoint.Index = uint32(i)
+			break
+		}
+	}
+	if contractOutPoint.Index == ^uint32(0) {
+		return nil, 0, errors.New("contract tx does not contain a P2SH contract payment")
+	}
+
+	refundAddress, err := getRawChangeAddress(c)
+	if err != nil {
+		return nil, 0, fmt.Errorf("getrawchangeaddress: %v", err)
+	}
+	refundOutScript, err := txscript.PayToAddrScript(refundAddress)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	pushes, err := txscript.ExtractAtomicSwapDataPushes(contract)
+	if err != nil {
+		// expected to only be called with good input
+		panic(err)
+	}
+
+	refundAddr, err := btcutil.NewAddressPubKeyHash(pushes.RefundHash160[:], chainParams)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	refundTx = wire.NewMsgTx(txVersion)
+	refundTx.LockTime = uint32(pushes.LockTime)
+	refundTx.AddTxOut(wire.NewTxOut(0, refundOutScript)) // amount set below
+	refundSize := estimateRefundSerializeSize(contract, refundTx.TxOut)
+	refundFee = txrules.FeeForSerializeSize(feePerKb, refundSize)
+	refundTx.TxOut[0].Value = contractTx.TxOut[contractOutPoint.Index].Value - int64(refundFee)
+	if txrules.IsDustOutput(refundTx.TxOut[0], minFeePerKb) {
+		return nil, 0, fmt.Errorf("refund output value of %v is dust", btcutil.Amount(refundTx.TxOut[0].Value))
+	}
+
+	txIn := wire.NewTxIn(&contractOutPoint, nil, nil)
+	txIn.Sequence = 0
+	refundTx.AddTxIn(txIn)
+
+	refundSig, refundPubKey, err := createSig(refundTx, 0, contract, refundAddr, c, contractTx.TxOut[contractOutPoint.Index].Value)
+	if err != nil {
+		return nil, 0, err
+	}
+	refundSigScript, err := refundP2SHContract(contract, refundSig, refundPubKey)
+	if err != nil {
+		return nil, 0, err
+	}
+	refundTx.TxIn[0].SignatureScript = refundSigScript
+
+	/*if verify {
```

Comment:
are you planning on implementing FORK_ID in the script engine so the signature verification can be done in the client before publishing?

Reply:
@jrick, it's pretty easy if I use cpacia's btcd, but it's not up to date, and it won't get btcsuite's btcd's bug fixes or other future commits. Should I fork it (already did) and keep it up to date using https://backstroke.co or should I use cpacia's abandoned btcd?